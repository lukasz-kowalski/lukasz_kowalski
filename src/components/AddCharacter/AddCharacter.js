import React from 'react';
import TextInput from '../ui/TextInput';
import Select from '../ui/Select';
import Radio from '../ui/Radio';

const radioData = [
  {
    id: 'character-male',
    value: 'male',
    label: 'Male'
  },
  {
    id: 'character-female',
    value: 'female',
    label: 'Female'
  },
  {
    id: 'character-na',
    value: 'n/a',
    label: 'n/a'
  }
]

class AddCharacter extends React.Component {
  constructor(props) {
    super(props);
    this['character-nameRef'] = React.createRef();
    this['character-speciesRef'] = React.createRef();
    this['character-genderRef'] = React.createRef();
  }
  state = {
    species: [],
    doValidate: false,
    isFormValid: false,
    inputs: {
      'character-name': {
        value: '',
        isValid: false
      },
      'character-species': {
        value: '',
        isValid: false
      },
      'character-gender': {
        value: '',
        isValid: false
      },
      'character-homeland': {
        value: '',
        isValid: true
      }
    }
  }

  componentDidMount() {
    this.fetchSpecies();
  }

  fetchSpecies = async () => {
    const response = await fetch('http://localhost:3000/species');
    const species = await response.json();
    this.setState({
      species
    });
  }

  handleSubmit = async event => {
    event.preventDefault();
    await this.setState({
      doValidate: true
    });
    if (!this.state.isFormValid) {
      this.setState({
        doValidate: false
      });
      this.focusFirstInvalidField();
    } else {
      try {
        const response = await this.postData();
        if (response.ok) {
          this.setState({
            doValidate: false
          });
          this.props.history.push('/');
        }
      } catch (err) {
        console.log(err);
      }
    }
  }

  handleInputValue = (id, value) => {
    this.setState({
      inputs: {
        ...this.state.inputs,
        [id]: {
          ...this.state.inputs[id],
          value
        }
      }
    });
  }

  validateForm = async (id, isInputValid) => {
    const keys = Object.keys(this.state.inputs);
    await this.setState({
      inputs: {
        ...this.state.inputs,
        [id]: {
          ...this.state.inputs[id],
          isValid: isInputValid
        }
      }
    });
    const allInputsValid = keys.every(key => this.state.inputs[key].isValid)
    await this.setState({
      isFormValid: allInputsValid,
    });
  }

  focusFirstInvalidField = () => {
    const keys = Object.keys(this.state.inputs);
    let firstInvalid;
    for (let key of keys) {
      if (!this.state.inputs[key].isValid) {
        firstInvalid = key
        this[`${firstInvalid}Ref`].current.focus();
        return;
      }
    }
  }

  createPayload = () => {
    const payload = {
      id: Math.floor(Math.random() * 10000),
      name: this.state.inputs['character-name'].value,
      species: this.state.inputs['character-species'].value,
      gender: this.state.inputs['character-gender'].value,
      homeworld: this.state.inputs['character-homeland'].value
    };
    return payload;
  }

  postData = () => {
    const payload = this.createPayload();
    return fetch('http://localhost:3000/characters', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload)
    });
  }

  render() {
    return (
      <main className="container">
        <h1>Add new character</h1>

        <form onSubmit={this.handleSubmit}>
          <TextInput
            id="character-name"
            label="Character name"
            isRequired={true}
            errorMsg="This field is required."
            placeholder="Enter name"
            doValidate={this.state.doValidate}
            handleInputValue={this.handleInputValue}
            validateForm={this.validateForm}
            reference={this['character-nameRef']}
          />
          <Select 
            id="character-species"
            label="Choose character species"
            options={this.state.species}
            isRequired={true}
            errorMsg="This field is required."
            doValidate={this.state.doValidate}
            handleInputValue={this.handleInputValue}
            validateForm={this.validateForm}
            reference={this['character-speciesRef']}
          />
          <Radio
            radioData={radioData}
            id="character-gender"
            label="Choose character gender"
            isRequired={true}
            errorMsg="This field is required."
            doValidate={this.state.doValidate}
            handleInputValue={this.handleInputValue}
            validateForm={this.validateForm}
            reference={this['character-genderRef']}
          />
          <TextInput
            id="character-homeland"
            label="Character homeland"
            placeholder="Enter homeland"
            handleInputValue={this.handleInputValue}
          />
          <button type="submit" className={`btn btn-primary ${this.state.doValidate && 'disabled'}`}>Submit</button>
        </form>
      </main>
    );
  };
};

export default AddCharacter;
