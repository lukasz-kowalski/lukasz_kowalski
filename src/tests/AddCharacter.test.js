import React from 'react';
import { shallow } from 'enzyme';
import AddCharacter from '../components/AddCharacter/AddCharacter';

let wrapper;
const mockMath = Object.create(global.Math);
mockMath.random = () => 1;
global.Math = mockMath;

beforeEach(() => {
  const species = ['aliens', 'martians'];
  fetch.mockResponse(JSON.stringify(species));
  wrapper = shallow(
  <AddCharacter />);
})

test('should successful render AddCharacter', () => {
  expect(wrapper).toMatchSnapshot();
})

test('should set character-name value to abc', () => {
  wrapper.instance().handleInputValue('character-name', 'abc');
  expect(wrapper.state().inputs['character-name'].value).toBe('abc');
})

test('should set response in species array', async () => {
  const species = ['aliens', 'martians'];
  await wrapper.instance().fetchSpecies();
  expect(wrapper.state('species')).toEqual(species);
})

test('should set isFormValid in state to false if any input is invalid', async () => {
  await wrapper.instance().validateForm('character-name', false);
  expect(wrapper.state('isFormValid')).toBe(false);
})

test('should set isFormValid in state to true if all inputs are valid', async () => {
  const inputs = ['character-name', 'character-species', 'character-gender'];
  await inputs.forEach(input => wrapper.instance().validateForm(input, true));
  expect(wrapper.state('isFormValid')).toBe(true);
})

test('should create expected payload', () => {
  const expectedPayload = {
    id: 10000,
    gender: 'male',
    homeworld: 'Mars',
    name: 'Last Man',
    species: 'martian'
  };
  const inputs = [
    { name: 'character-name', value: 'Last Man' },
    { name: 'character-species', value: 'martian' },
    { name: 'character-gender', value: 'male' },
    { name: 'character-homeland', value: 'Mars' }
  ];

  inputs.forEach(input => wrapper.instance().handleInputValue(input.name, input.value));
  expect(wrapper.instance().createPayload()).toEqual(expectedPayload);
})
