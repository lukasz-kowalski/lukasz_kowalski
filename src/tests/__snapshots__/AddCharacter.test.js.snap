// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should successful render AddCharacter 1`] = `
<main
  className="container"
>
  <h1>
    Add new character
  </h1>
  <form
    onSubmit={[Function]}
  >
    <WithHandlers
      doValidate={false}
      errorMsg="This field is required."
      handleInputValue={[Function]}
      id="character-name"
      isRequired={true}
      label="Character name"
      placeholder="Enter name"
      reference={
        Object {
          "current": null,
        }
      }
      validateForm={[Function]}
    />
    <WithHandlers
      doValidate={false}
      errorMsg="This field is required."
      handleInputValue={[Function]}
      id="character-species"
      isRequired={true}
      label="Choose character species"
      options={Array []}
      reference={
        Object {
          "current": null,
        }
      }
      validateForm={[Function]}
    />
    <WithHandlers
      doValidate={false}
      errorMsg="This field is required."
      handleInputValue={[Function]}
      id="character-gender"
      isRequired={true}
      label="Choose character gender"
      radioData={
        Array [
          Object {
            "id": "character-male",
            "label": "Male",
            "value": "male",
          },
          Object {
            "id": "character-female",
            "label": "Female",
            "value": "female",
          },
          Object {
            "id": "character-na",
            "label": "n/a",
            "value": "n/a",
          },
        ]
      }
      reference={
        Object {
          "current": null,
        }
      }
      validateForm={[Function]}
    />
    <WithHandlers
      handleInputValue={[Function]}
      id="character-homeland"
      label="Character homeland"
      placeholder="Enter homeland"
    />
    <button
      className="btn btn-primary false"
      type="submit"
    >
      Submit
    </button>
  </form>
</main>
`;
